fn/mvar.go

```diff
@@ -0,0 +1,180 @@
+package fn
+
+import (
+	"sync/atomic"
+)
+
+// MVar[A any] is a structure that is designed to store a single value in an API
+// that dispenses with data races. Think of it as a box for a value.
+//
+// It has two states: full and empty.
+//
+// It supports two operations: take and put.
+//
+// The state transition rules are as follows:
+//  1. put while full blocks.
+//  2. put while empty sets.
+//  3. take while full resets.
+//  4. take while empty blocks.
+//  5. read while full nops.
+//  6. read while empty blocks.
+type MVar[A any] struct {
+	// current is an immediately available copy of whatever is inside the
+	// value channel that is served to readers. It is updated whenever a
+	// change to the value channel is successful.
+	current *atomic.Pointer[A]
+	// readers is used to wake all blocked readers when a new value is
+	// written.
+	readers chan chan A
+
+	// takers is used to wake a single taker when a new value is written.
+	takers chan chan A
+
+	// value is a bounded channel of size 1 that represents the core state
+	// oof the channel.
+	value chan A
```

Comment:
Alternatively could use the `CompareAndSwap` API (https://pkg.go.dev/sync/atomic#Pointer.CompareAndSwap) which may simplify the code a bit.

Reply:
I tried this in an earlier version of this code but the issue was that I still needed a blocking facility. I tried using `sync.Cond` to accomplish this but the guarantees provided by Cond.Broadcast did not fit the bill I was looking for. Additionally the code that used Conds was far more complex than this implementation.

Reply:
I figured out how to use CAS+Conds and solve my problem of racey reads/takes